Declare pip dependency for cli extras

jonbiemond opened the pull request:
The `pipdeptree` package includes references to `pip` but does not explicitly require it.
To ensure `pip` is installed in a users environment, include it as a dependency.

Fixes #2203

jonbiemond commented:
pip's versioning scheme is `year.quarter.patch`. I arbitrarily chose to pin dependencies at `>=24.0.0` or the earliest release of 2024.

burnash commented:
Hey @jonbiemond, thank you for the fix and sorry for the delayed response. Could you please merge or rebase the dev branch and regenerate `poetry.lock` to fix conflicts?

jonbiemond commented:
Hi @burnash, no worries! I've rebased and regenerated the lock file.

There are also two tests failing in CI. Now that pip is explicitly declared as an optional dependency, `poetry install` will remove it if it exists and when no extras are specified. These two tests rely on pip, which is why they now fail.

I've added a commit to include pip in the dev dependencies as well. This should resolve the failing tests. Alternatively all dependencies could be installed in CI with the `--all-extras` flag.

sh-rp commented:
lgtm, I'll merge it @burnash
